File: figures/src/python/classes/volcano_pathway_compound.py
#!/usr/bin/env python3
"""Class-based Python renderer for volcano plus pathway compound figures."""

from __future__ import annotations

import csv
import math
from pathlib import Path
from typing import Any

from matplotlib.figure import Figure

from figures.src.python.common import (
    REPO_ROOT,
    add_panel_labels,
    apply_publication_layout,
    configure_matplotlib,
    export_csv,
    mm_to_inches,
    project_resources,
    resolve_font_stack,
    source_data_mapping,
    validate_common_contract,
    write_manifest,
)


def _read_gene_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        rows: list[dict[str, Any]] = []
        for row in reader:
            padj = float(row["padj"])
            rows.append(
                {
                    "gene": row["gene"],
                    "log2_fc": float(row["log2_fc"]),
                    "padj": padj,
                    "neg_log10_padj": -math.log10(max(padj, 1e-300)),
                    "highlight_label": row["highlight_label"],
                }
            )
        return rows


def _read_pathway_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "pathway": row["pathway"],
                "nes": float(row["nes"]),
                "fdr": float(row["fdr"]),
                "direction": row["direction"],
                "highlight_order": int(row["highlight_order"]),
            }
            for row in reader
        ]


def build_source_data(
    spec: dict[str, Any],
    gene_rows: list[dict[str, Any]],
    pathway_rows: list[dict[str, Any]],
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    classified: list[dict[str, Any]] = []
    for row in gene_rows:
        if float(row["log2_fc"]) >= 1.0 and float(row["padj"]) <= 0.05:
            category = "up_significant"
        elif float(row["log2_fc"]) <= -1.0 and float(row["padj"]) <= 0.05:
            category = "down_significant"
        else:
            category = "background"
        classified.append({**row, "significance_category": category})

    ordered_pathways = [
        {**row, "fdr_label": f"FDR {row['fdr']:.1e}"}
        for row in sorted(pathway_rows, key=lambda item: item["nes"], reverse=True)
    ]
    source_outputs = source_data_mapping(spec)
    export_csv(
        REPO_ROOT / source_outputs["a"],
        classified,
        [
            "gene",
            "log2_fc",
            "padj",
            "neg_log10_padj",
            "highlight_label",
            "significance_category",
        ],
    )
    export_csv(
        REPO_ROOT / source_outputs["b"],
        ordered_pathways,
        [
            "pathway",
            "nes",
            "fdr",
            "direction",
            "highlight_order",
            "fdr_label",
        ],
    )
    return classified, ordered_pathways


def create_figure(spec_path: Path) -> Figure:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    gene_rows = _read_gene_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    pathway_rows = _read_pathway_rows(REPO_ROOT / str(spec["data_inputs"][1]))
    panel_a_rows, panel_b_rows = build_source_data(spec, gene_rows, pathway_rows)
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)

    import matplotlib.pyplot as plt

    fig, axes = plt.subplots(
        1,
        2,
        figsize=(
            mm_to_inches(float(spec["size"]["width_mm"])),
            mm_to_inches(float(spec["size"]["height_mm"])),
        ),
        constrained_layout=bool(theme["layout"]["constrained"]),
    )
    apply_publication_layout(fig, theme)

    colors = {
        "background": theme["palette"]["neutral"][2],
        "up_significant": theme["palette"]["categorical"][1],
        "down_significant": theme["palette"]["categorical"][0],
        "up": theme["palette"]["categorical"][1],
        "down": theme["palette"]["categorical"][0],
    }

    panel_a = axes[0]
    for category in ("background", "down_significant", "up_significant"):
        subset = [row for row in panel_a_rows if row["significance_category"] == category]
        if subset:
            panel_a.scatter(
                [float(row["log2_fc"]) for row in subset],
                [float(row["neg_log10_padj"]) for row in subset],
                s=18 if category == "background" else 28,
                color=colors[category],
                alpha=0.55 if category == "background" else 0.92,
                edgecolor="none",
            )
    threshold_y = -math.log10(0.05)
    panel_a.axvline(-1.0, color=theme["palette"]["neutral"][1], linestyle="--", linewidth=0.9)
    panel_a.axvline(1.0, color=theme["palette"]["neutral"][1], linestyle="--", linewidth=0.9)
    panel_a.axhline(threshold_y, color=theme["palette"]["neutral"][1], linestyle="--", linewidth=0.9)
    label_offsets = {
        "CXCL10": (0.18, 0.36),
        "IFIT1": (0.16, 0.12),
        "MX1": (0.16, -0.12),
        "MKI67": (-0.18, 0.34),
        "TOP2A": (-0.18, 0.10),
        "CDK1": (-0.18, -0.14),
    }
    for row in [row for row in panel_a_rows if row["highlight_label"] == "yes"]:
        dx, dy = label_offsets.get(str(row["gene"]), (0.14, 0.14))
        x = float(row["log2_fc"])
        y = float(row["neg_log10_padj"])
        panel_a.annotate(
            str(row["gene"]),
            xy=(x, y),
            xytext=(x + dx, y + dy),
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][0],
            ha="left" if dx >= 0 else "right",
            va="center",
            arrowprops={
                "arrowstyle": "-",
                "color": theme["palette"]["neutral"][1],
                "linewidth": 0.7,
            },
        )
    max_abs_fc = max(abs(float(row["log2_fc"])) for row in panel_a_rows)
    max_y = max(float(row["neg_log10_padj"]) for row in panel_a_rows)
    panel_a.set_xlim(-(max_abs_fc + 0.8), max_abs_fc + 0.8)
    panel_a.set_ylim(0, max_y + 1.45)
    panel_a.set_title("Differential expression volcano", loc="left", pad=2)
    panel_a.set_xlabel("log2 fold change")
    panel_a.set_ylabel("-log10 adjusted P")
    panel_a.spines["top"].set_visible(False)
    panel_a.spines["right"].set_visible(False)

    panel_b = axes[1]
    y_positions = list(range(len(panel_b_rows)))
    panel_b.barh(
        y_positions,
        [float(row["nes"]) for row in panel_b_rows],
        color=[colors[str(row["direction"])] for row in panel_b_rows],
        alpha=0.9,
        edgecolor=theme["palette"]["neutral"][0],
        linewidth=theme["strokes"]["errorbar_line_width_pt"],
    )
    panel_b.axvline(0.0, color=theme["palette"]["neutral"][1], linewidth=0.9)
    panel_b.set_yticks(y_positions)
    panel_b.set_yticklabels([str(row["pathway"]) for row in panel_b_rows])
    panel_b.invert_yaxis()
    x_extent = max(abs(float(row["nes"])) for row in panel_b_rows) + 0.65
    panel_b.set_xlim(-x_extent, x_extent)
    for index, row in enumerate(panel_b_rows):
        nes = float(row["nes"])
        panel_b.text(
            nes + (0.12 if nes >= 0 else -0.12),
            index,
            str(row["fdr_label"]),
            ha="left" if nes >= 0 else "right",
            va="center",
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][1],
        )
    panel_b.set_title("Pathway enrichment summary", loc="left", pad=2)
    panel_b.set_xlabel("Normalized enrichment score")
    panel_b.spines["top"].set_visible(False)
    panel_b.spines["right"].set_visible(False)

    add_panel_labels(spec, theme, [panel_a, panel_b])
    return fig


def build_figure(spec_path: Path) -> dict[str, str]:
    spec, theme, font_policy, profile, _ = project_resources(spec_path)
    gene_rows = _read_gene_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    resolved_font = resolve_font_stack(font_policy)
    fig = create_figure(spec_path)
    output_dir = REPO_ROOT / str(spec["renderers"]["python"]["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)
    stem = str(spec["figure_id"])
    svg_path = output_dir / f"{stem}.svg"
    pdf_path = output_dir / f"{stem}.pdf"
    png_path = output_dir / f"{stem}.png"
    fig.savefig(svg_path, format="svg")
    fig.savefig(pdf_path, format="pdf")
    fig.savefig(png_path, format="png", dpi=int(profile["preview_dpi"]))
    outputs = {
        "svg": str(svg_path.relative_to(REPO_ROOT)),
        "pdf": str(pdf_path.relative_to(REPO_ROOT)),
        "png": str(png_path.relative_to(REPO_ROOT)),
    }
    write_manifest(
        spec,
        profile,
        resolved_font,
        "python",
        spec_path,
        outputs,
        [
            "threshold_guides_for_significance",
            "selective_gene_labels_for_extreme_points",
            "de_emphasized_nonsignificant_points",
            "signed_enrichment_panel",
            "vector_first_export",
        ],
        {
            "panel_count": len(spec.get("panels", [])),
            "highlight_label_count": sum(
                1 for row in gene_rows if str(row["highlight_label"]) == "yes"
            ),
            "threshold_line_count": 3,
            "reference_line_count": 1,
        },
    )
    return outputs


if __name__ == "__main__":
    raise SystemExit("Use figures/src/python/run_class_renderer.py")

File: figures/src/r/classes/volcano_pathway_compound.R
repo_root <- getOption("manuscript.figure.repo_root")
if (is.null(repo_root)) {
  stop("manuscript.figure.repo_root option is required")
}
source(file.path(repo_root, "figures", "src", "r", "common.R"))

create_plot <- function(spec_path) {
  spec <- load_yaml(spec_path)
  theme <- load_yaml(file.path(repo_root, "figures/config/project_theme.yml"))
  font_policy <- load_yaml(file.path(repo_root, "figures/config/font_policy.yml"))
  profiles <- load_yaml(file.path(repo_root, "figures/config/venue_profiles.yml"))
  registry <- load_yaml(file.path(repo_root, "figures/registry/classes.yml"))$classes
  profile <- profiles$profiles[[spec$target_profile]]
  class_entry <- registry[[spec$class_id]]
  validate_common_contract(spec, theme, font_policy, profile, class_entry)
  resolved_font <- resolve_font(font_policy)
  plot_theme <- build_theme(theme, resolved_font)

  gene_rows <- read.csv(file.path(repo_root, spec$data_inputs[[1]]), stringsAsFactors = FALSE, check.names = FALSE)
  pathway_rows <- read.csv(file.path(repo_root, spec$data_inputs[[2]]), stringsAsFactors = FALSE, check.names = FALSE)
  gene_rows$neg_log10_padj <- -log10(pmax(gene_rows$padj, 1e-300))
  gene_rows$significance_category <- ifelse(
    gene_rows$log2_fc >= 1 & gene_rows$padj <= 0.05,
    "up_significant",
    ifelse(gene_rows$log2_fc <= -1 & gene_rows$padj <= 0.05, "down_significant", "background")
  )
  pathway_rows$fdr_label <- sprintf("FDR %.1e", pathway_rows$fdr)
  pathway_rows <- pathway_rows[order(pathway_rows$nes, decreasing = TRUE), ]
  pathway_rows$pathway <- factor(pathway_rows$pathway, levels = rev(pathway_rows$pathway))

  mapping <- source_data_mapping(spec)
  write_source_csv(
    file.path(repo_root, mapping$a),
    gene_rows,
    c(
      "gene",
      "log2_fc",
      "padj",
      "neg_log10_padj",
      "highlight_label",
      "significance_category"
    )
  )
  write_source_csv(
    file.path(repo_root, mapping$b),
    pathway_rows,
    c("pathway", "nes", "fdr", "direction", "highlight_order", "fdr_label")
  )

  label_positions <- data.frame(
    gene = c("CXCL10", "IFIT1", "MX1", "MKI67", "TOP2A", "CDK1"),
    label_dx = c(0.18, 0.16, 0.16, -0.18, -0.18, -0.18),
    label_dy = c(0.36, 0.12, -0.12, 0.34, 0.10, -0.14),
    stringsAsFactors = FALSE
  )
  gene_labels <- merge(gene_rows[gene_rows$highlight_label == "yes", ], label_positions, by = "gene", sort = FALSE)
  gene_labels$label_x <- gene_labels$log2_fc + gene_labels$label_dx
  gene_labels$label_y <- gene_labels$neg_log10_padj + gene_labels$label_dy
  threshold_y <- -log10(0.05)

  plot_a <- ggplot2::ggplot(gene_rows, ggplot2::aes(x = log2_fc, y = neg_log10_padj)) +
    ggplot2::geom_vline(xintercept = c(-1, 1), linewidth = 0.4, linetype = "dashed", color = theme$palette$neutral[[2]]) +
    ggplot2::geom_hline(yintercept = threshold_y, linewidth = 0.4, linetype = "dashed", color = theme$palette$neutral[[2]]) +
    ggplot2::geom_point(ggplot2::aes(color = significance_category, alpha = significance_category), size = 1.8) +
    ggplot2::scale_color_manual(values = c("background" = theme$palette$neutral[[3]], "down_significant" = theme$palette$categorical[[1]], "up_significant" = theme$palette$categorical[[2]])) +
    ggplot2::scale_alpha_manual(values = c("background" = 0.55, "down_significant" = 0.92, "up_significant" = 0.92)) +
    ggplot2::geom_segment(
      data = gene_labels,
      ggplot2::aes(x = log2_fc, y = neg_log10_padj, xend = label_x, yend = label_y),
      inherit.aes = FALSE,
      linewidth = 0.35,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::geom_text(
      data = gene_labels,
      ggplot2::aes(x = label_x, y = label_y, label = gene, hjust = ifelse(label_dx >= 0, 0, 1)),
      inherit.aes = FALSE,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::labs(title = "Differential expression volcano", x = "log2 fold change", y = "-log10 adjusted P") +
    ggplot2::coord_cartesian(xlim = c(-max(abs(gene_rows$log2_fc)) - 0.8, max(abs(gene_rows$log2_fc)) + 0.8), ylim = c(0, max(gene_rows$neg_log10_padj) + 1.1)) +
    plot_theme

  plot_b <- ggplot2::ggplot(pathway_rows, ggplot2::aes(x = nes, y = pathway, fill = direction)) +
    ggplot2::geom_vline(xintercept = 0, linewidth = 0.4, color = theme$palette$neutral[[2]]) +
    ggplot2::geom_col(width = 0.7, alpha = 0.9, color = theme$palette$neutral[[1]], linewidth = 0.35) +
    ggplot2::scale_fill_manual(values = c("down" = theme$palette$categorical[[1]], "up" = theme$palette$categorical[[2]])) +
    ggplot2::geom_text(ggplot2::aes(label = fdr_label, x = ifelse(nes >= 0, nes + 0.12, nes - 0.12), hjust = ifelse(nes >= 0, 0, 1)), size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845, color = theme$palette$neutral[[2]]) +
    ggplot2::labs(title = "Pathway enrichment summary", x = "Normalized enrichment score", y = NULL) +
    ggplot2::coord_cartesian(xlim = c(-max(abs(pathway_rows$nes)) - 0.65, max(abs(pathway_rows$nes)) + 0.65)) +
    plot_theme

  figure_plot <- (plot_a + plot_b + patchwork::plot_layout(ncol = 2)) +
    patchwork::plot_annotation(tag_levels = "a") &
    ggplot2::theme(
      plot.tag = ggplot2::element_text(
        size = as.numeric(theme$panel_labels$font_size_pt),
        face = theme$panel_labels$font_weight,
        family = resolved_font$family
      )
    )
  list(
    plot = figure_plot,
    spec = spec,
    theme = theme,
    profile = profile,
    resolved_font = resolved_font,
    gene_rows = gene_rows
  )
}

build_figure <- function(spec_path) {
  payload <- create_plot(spec_path)
  spec <- payload$spec
  theme <- payload$theme
  profile <- payload$profile
  resolved_font <- payload$resolved_font
  figure_plot <- payload$plot
  gene_rows <- payload$gene_rows
  output_dir <- file.path(repo_root, spec$renderers$r$output_dir)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  stem <- spec$figure_id
  svg_path <- file.path(output_dir, sprintf("%s.svg", stem))
  pdf_path <- file.path(output_dir, sprintf("%s.pdf", stem))
  png_path <- file.path(output_dir, sprintf("%s.png", stem))
  width_in <- mm_to_inches(as.numeric(spec$size$width_mm))
  height_in <- mm_to_inches(as.numeric(spec$size$height_mm))

  svglite::svglite(file = svg_path, width = width_in, height = height_in, bg = theme$export_defaults$background, system_fonts = list(sans = resolved_font$family))
  print(figure_plot)
  grDevices::dev.off()
  grDevices::cairo_pdf(file = pdf_path, width = width_in, height = height_in, bg = theme$export_defaults$background, family = resolved_font$family)
  print(figure_plot)
  grDevices::dev.off()
  ragg::agg_png(filename = png_path, width = width_in, height = height_in, units = "in", res = as.numeric(profile$preview_dpi), background = theme$export_defaults$background)
  print(figure_plot)
  grDevices::dev.off()

  outputs <- list(
    svg = sub(paste0("^", repo_root, "/"), "", svg_path),
    pdf = sub(paste0("^", repo_root, "/"), "", pdf_path),
    png = sub(paste0("^", repo_root, "/"), "", png_path)
  )
  write_manifest(
    repo_root,
    spec,
    profile,
    resolved_font,
    spec_path,
    "r",
    outputs,
    list(
      "threshold_guides_for_significance",
      "selective_gene_labels_for_extreme_points",
      "de_emphasized_nonsignificant_points",
      "signed_enrichment_panel",
      "vector_first_export"
    ),
    list(
      panel_count = length(spec$panels),
      highlight_label_count = sum(gene_rows$highlight_label == "yes"),
      threshold_line_count = 3,
      reference_line_count = 1
    )
  )
  outputs
}
